modules/memberships/class-jetpack-memberships.php

```diff
@@ -227,21 +227,81 @@ public function allow_sync_post_meta( $post_meta ) {
 	public function return_meta( $map ) {
 		return $map['meta'];
 	}
+
 	/**
-	 * Callback that parses the membership purchase shortcode.
+	 * Renders a preview of the Recurring Payment button, which is not hooked
+	 * up to the subscription url. Used to preview the block on the frontend
+	 * for admin users when Stripe has not been connected.
 	 *
-	 * @param array  $attrs - attributes in the shortcode. `id` here is the CPT id of the plan.
+	 * @param array  $attrs - attributes in the shortcode.
 	 * @param string $content - Recurring Payment block content.
 	 *
 	 * @return string|void
 	 */
-	public function render_button( $attrs, $content = null ) {
+	public function render_button_preview( $attrs, $content = null ) {
+		if ( ! empty( $content ) ) {
+			$block_id = esc_attr( wp_unique_id( 'recurring-payments-block-' ) );
+			$content  = str_replace( 'recurring-payments-id', $block_id, $content );
+			return $content;
+		}
+		return $this->deprecated_render_button_v1( $attrs, null );
+	}
+
+	/**
+	 * Determines whether the button preview should be rendered. Returns true
+	 * if the user is an admin, the button is not configured correctly
+	 * (because it requires a plan upgrade or Stripe connection), and the
+	 * button is a child of a Premium Content block.
+	 *
+	 * @param WP_Block $block Recurring Payments block instance.
+	 *
+	 * @return boolean
+	 */
+	public function should_render_button_preview( $block ) {
+		$is_admin                   = $this->user_can_edit();
+		$requires_stripe_connection = ! $this->get_connected_account_id();
+
+		$requires_upgrade = false;
+		if ( defined( 'IS_WPCOM' ) && IS_WPCOM && function_exists( 'has_any_blog_stickers' ) ) {
+			$site_id          = get_current_blog_id();
+			$requires_upgrade = has_any_blog_stickers( array( 'personal-plan', 'premium-plan', 'business-plan', 'ecommerce-plan' ), $site_id );
```

Comment:
We do things a bit differently elsewhere? 
https://github.com/Automattic/jetpack/blob/9222df3c888953e6f0d867e4c4d0e0e47c1ad5f2/class.jetpack-gutenberg.php#L1017-L1021

Should we / Could we take a similar approach here?

We also have `is_enabled_jetpack_recurring_payments` in this file. Maybe that's an opportunity to merge things a bit?

Reply:
I'm not very familiar with this code, but when I inspect `$features_data` on a site which I know supports recurring payments (WPCOM site with a premium plan), I don't see any entry for recurring-payments.

`is_enabled_jetpack_recurring_payments` does almost what we want but not quite -- rather than checking whether the block is supported by the site's plan, it checks whether it ought to be enabled. This is true for Jetpack sites when the upgrade nudge is enabled, even if the block is not supported.

I agree that's a good opportunity to merge things a bit; I'll try to refactor that function slightly.